### PR TITLE
Ipmcvg 1412 improve alert system

### PR DESCRIPTION
**Pull request by nicodav**
Implement the sse protocol for send new Alert to the UI with a new eccp file : admin_sse.ecpp.
Base url : <host>/api/v1/sse?datacenter=xxx
Parameter : datacenter : the id of the datacenter to manage. 

Functionnality : 
-Must have a valid connection to get a sse connection.
-Subscribe to the ALERT stream and when receive an alert, send this alert via sse if the asset of the alert is a part of the datacenter.
-If no alert is send to UI after waiting 10 seconds, a heartbeat message is send to control the UI is still connected.
-Every minute, a message with the remaining time of the session is send.
-If the session is revoked or expired, the  sse connection is shutdown.
-If any notification (alert or heartbeat or session time) failed when send, the sse connection is shutdown
-The sse timout connection is the half of the tntnet request timeout

See https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events for sse specifications and recommandations

### CODE REVIEW

File: src/web/src/admin_sse.ecpp [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    if (!dc.empty ())
+    {
+        if (!is_ok_name (dc.c_str ()))
+        {
+            http_die ("request-param-bad", dc.c_str ());
+        }
+    } 
+
+    checked_dc = dc;
+    std::string checked_state = "ALL";
```
Comment (geraldguillaume) [inline src/web/src/admin_sse.ecpp]: checked_state is never used

File: src/web/src/admin_sse.ecpp [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    log_debug ("Asset element id = '%" PRIu32 "'.", element_id);
+
+    if (!checked_dc.empty ()) {
+        auto asset_element = persist::select_asset_element_web_byId (connection, element_id);
+        if (asset_element.status != 1)
+            http_die ("internal-error", asset_element.msg.c_str ());
+        log_debug ("Asset element name = '%s'.", asset_element.item.name.c_str ());
+        desired_elements.emplace (std::make_pair (asset_element.item.name, 5));
+
+        if (checked_recursive.compare ("true") == 0) {
```
Comment (geraldguillaume) [inline src/web/src/admin_sse.ecpp]: checked_recursive is initialized by "true" and never re-affected.
This variable looks useless

File: src/web/src/tokens.cc [hunk 3]
```diff
@@ -253,11 +253,15 @@ BiosProfile tokens::verify_token(const std::string token, long int* uid, long in
     if (gid)
         *gid = l_gid;
 
+    time_t now = mono_time (NULL);
```
Comment (geraldguillaume) [inline src/web/src/tokens.cc]: reuse now in the condition line bellow "if(mono_time(NULL).."

File: src/web/src/tokens.cc [hunk 4]
```diff
@@ -253,11 +253,15 @@ BiosProfile tokens::verify_token(const std::string token, long int* uid, long in
...
     if (mono_time (NULL) > tme) {
         log_info ("verify_token: expired token for uid/gid %ld/%ld, authentication failed!", l_uid, l_gid);
         return BiosProfile::Anonymous;
     }
-
+    *expInSec = tme - now;
+    log_debug ("verify_token: expInSec: %ld",*expInSec);
```
Comment (geraldguillaume) [inline src/web/src/tokens.cc]: This debug could really spam the system. I recommend to remove it

File: src/web/src/tokens.cc [hunk 5]
```diff
@@ -253,11 +253,15 @@ BiosProfile tokens::verify_token(const std::string token, long int* uid, long in
...
+
```
Comment (geraldguillaume) [inline src/web/src/tokens.cc]: two empty line is too much

File: src/web/src/admin_sse.ecpp [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+    if (!dc.empty ())
+    {
+        if (!is_ok_name (dc.c_str ()))
+        {
+            http_die ("request-param-bad", dc.c_str ());
+        }
+    } 
+
+    checked_dc = dc;
```
[the remark this answers is not in the document]
Comment (nicodav) [reply src/web/src/admin_sse.ecpp]: Remove checked_dc variable

**Pull request merged**